File: legacy/src/main/rs/emulate/legacy/version/MapIndexDecoder.java
package rs.emulate.legacy.version;

import java.io.FileNotFoundException;

import rs.emulate.legacy.archive.Archive;
import rs.emulate.shared.util.DataBuffer;

/**
 * Decoder for the version lists "map_index" entry.
 *
 * @author sfix
 * @author Major
 */
public class MapIndexDecoder {

	/**
	 * The name of the archive entry containing the map index data.
	 */
	private static final String ENTRY_NAME = "map_index";

	/**
	 * The DataBuffer containing the data.
	 */
	private final DataBuffer data;

	/**
	 * Creates the MapIndexDecoder.
	 *
	 * @param archive The version list {@link Archive}.
	 * @throws FileNotFoundException If the {@code map_index} entry could not be found.
	 */
	public MapIndexDecoder(Archive archive) throws FileNotFoundException {
		data = archive.getEntry(ENTRY_NAME).getBuffer().asReadOnlyBuffer();
	}

	/**
	 * Decodes the contents of the {@code map_index} entry into a {@link MapIndex}.
	 *
	 * @return The MapIndex.
	 */
	public MapIndex decode() {
		int count = data.remaining() / (3 * Short.BYTES + Byte.BYTES);

		int[] areas = new int[count];
		int[] landscapes = new int[count];
		int[] maps = new int[count];
		boolean[] members = new boolean[count];

		for (int index = 0; index < count; index++) {
			areas[index] = data.getUnsignedShort();
			maps[index] = data.getUnsignedShort();
			landscapes[index] = data.getUnsignedShort();
			members[index] = data.getBoolean();
		}

		return new MapIndex(areas, landscapes, maps, members);
	}

}

File: legacy/src/main/rs/emulate/legacy/version/MapIndexDecoder.kt
package rs.emulate.legacy.version

import rs.emulate.legacy.archive.Archive
import rs.emulate.shared.util.DataBuffer

import java.io.FileNotFoundException

/**
 * Decoder for the version lists "map_index" entry.
 *
 * @param archive The version list [Archive].
 * @throws FileNotFoundException If the `map_index` entry could not be found.
 */
class MapIndexDecoder constructor(archive: Archive) {

    private val data: DataBuffer = archive.getEntry(MapIndex.ENTRY_NAME).buffer.asReadOnlyBuffer()

    /**
     * Decodes the contents of the `map_index` entry into a [MapIndex].
     */
    fun decode(): MapIndex {
        val count = data.remaining() / (3 * java.lang.Short.BYTES + java.lang.Byte.BYTES)

        val areas = IntArray(count)
        val maps = IntArray(count)
        val objects = IntArray(count)
        val members = BooleanArray(count)

        for (index in 0 until count) {
            areas[index] = data.getUnsignedShort()
            maps[index] = data.getUnsignedShort()
            objects[index] = data.getUnsignedShort()
            members[index] = data.getBoolean()
        }

        return MapIndex(areas, objects, maps, members)
    }

}
